light/beacon/api/light_api.go

```diff
@@ -0,0 +1,372 @@
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"math/big"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	cbeacon "github.com/ethereum/go-ethereum/core/beacon"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/light/beacon"
+)
+
+// BeaconLightApi requests light client information from a beacon node REST API.
+// Note: all required API endpoints are currently only implemented by Lodestar.
+type BeaconLightApi struct {
+	url           string
+	client        *http.Client
+	customHeaders map[string]string
+}
+
+func NewBeaconLightApi(url string, customHeaders map[string]string) *BeaconLightApi {
+	return &BeaconLightApi{
+		url: url,
+		client: &http.Client{
+			Timeout: time.Second * 10,
+		},
+		customHeaders: customHeaders,
+	}
+}
+
+func (api *BeaconLightApi) httpGet(path string) ([]byte, error) {
+	req, err := http.NewRequest("GET", api.url+path, nil)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range api.customHeaders {
+		req.Header.Set(k, v)
+	}
+	resp, err := api.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return io.ReadAll(resp.Body)
+}
+
+// Header defines a beacon header and supports JSON encoding according to the standard beacon API format
+type jsonHeader struct {
+	Slot          common.Decimal `json:"slot"`
+	ProposerIndex common.Decimal `json:"proposer_index"`
+	ParentRoot    common.Hash    `json:"parent_root"`
+	StateRoot     common.Hash    `json:"state_root"`
+	BodyRoot      common.Hash    `json:"body_root"`
+}
+
+func (h *jsonHeader) header() beacon.Header {
+	return beacon.Header{
+		Slot:          uint64(h.Slot),
+		ProposerIndex: uint(h.ProposerIndex),
+		ParentRoot:    h.ParentRoot,
+		StateRoot:     h.StateRoot,
+		BodyRoot:      h.BodyRoot,
+	}
+}
+
+// GetBestUpdateAndCommittee fetches and validates LightClientUpdate for given period and full serialized
+// committee for the next period (committee root hash equals update.NextSyncCommitteeRoot).
+// Note that the results are validated but the update signature should be verified by the caller as its
+// validity depends on the update chain.
+func (api *BeaconLightApi) GetBestUpdateAndCommittee(period uint64) (beacon.LightClientUpdate, []byte, error) {
+	resp, err := api.httpGet("/eth/v1/beacon/light_client/updates?start_period=" + strconv.Itoa(int(period)) + "&count=1")
+	if err != nil {
+		return beacon.LightClientUpdate{}, nil, err
+	}
+
+	type committeeUpdate struct {
+		Header                  jsonHeader          `json:"attested_header"`
+		NextSyncCommittee       syncCommitteeJson   `json:"next_sync_committee"`
+		NextSyncCommitteeBranch beacon.MerkleValues `json:"next_sync_committee_branch"`
+		FinalizedHeader         jsonHeader          `json:"finalized_header"`
+		FinalityBranch          beacon.MerkleValues `json:"finality_branch"`
+		Aggregate               syncAggregate       `json:"sync_aggregate"`
```

Comment:
Please add SignatureSlot according to the latest spec

Reply:
Indeed, maybe it's better to decode and check it explicitly. Though later the signature period is assumed to be the same and the sig check fails if it is not, but now we catch it earlier and return a more informative error.